core: refactor RepeatablePath into separate files and add Path

jcollie opened the pull request:
Slims down `Config.zig` and makes some of the code reusable in Path.

jcollie committed:
core: all paths referenced from the CLI must be expanded

pluiedev commented on src/config/Config.zig:
```diff
@@ -2828,12 +2829,15 @@ fn expandPaths(self: *Config, base: []const u8) !void {
+                    &self._diagnostics,
+                );
+            },
+            inline else => {},
```
Does it need to be inline here? We don't do anything with the type anyway

pluiedev commented on src/config/NonRepeatablePath.zig:
```diff
@@ -0,0 +1,180 @@
+
+/// Deep copy of the struct. Required by Config.
+pub fn clone(self: *const NonRepeatablePath, alloc: Allocator) Allocator.Error!NonRepeatablePath {
+    if (self.value) |value| {
+        return .{
+            .value = try value.clone(alloc),
+        };
+    }
+    return .{
+        .value = null,
+    };
```
Simplify

```suggestion
    return .{
        .value = if (self.value) |v| v.clone(alloc) else null,
    };
```

jcollie replied:
Fixed.

pluiedev reviewed (changes requested):
NonRepeatablePath is kind of a funny name for an optional path. Is there anything stopping us from using ?Path directly?

jcollie commented:
> NonRepeatablePath is kind of a funny name for an optional path. Is there anything stopping us from using ?Path directly?

`Path` is a normal `union`, I don't think that the rest of the config has the machinery for dealing with non-packed unions.

mitchellh reviewed (changes requested):
Since they're all path-related, I suggest we put all three into `src/config/path.zig` instead.

jcollie committed:
core: move RepeatablePath to separate file and enable Path as config type

Slim down Config.zig by moving RepeatablePath to a separate file and
enable the use of Path as it's own config type.

jcollie commented:
OK, switched to just exposing `Path` as a config type and merged everything into `src/config/path.zig`.

pluiedev commented on src/config/path.zig:
```diff
@@ -0,0 +1,470 @@
+/// expanded relative to the path of the config file (or the home directory).
+pub const Path = union(enum) {
+    /// Path is unset.
+    null,
+
```
Why is `null` a variant? I'm pretty sure our config mechanism supports just setting a field to `?Path`

jcollie replied:
Fixed.
